### DISCUSSION TITLE
Custom Tile Size for custom WMS TileLayers in DeckGL

### DISCUSSION THREAD

**Discussion by lm93547** [visgl/deck.gl#6450] [Q&A] [answered]
Hi Folks, love the library. 
I am pulling some data from a WMS server which shows a map of the geological data of the United Kingdom. The data projection is EPSG27700 (UK Grid). I have got the layer rendering correctly using the correct URL params shown here: 
[https://map.bgs.ac.uk/arcgis/rest/services/GeoIndex_Onshore/hydrogeology/MapServer//export?dpi=96&transparent=true&format=png&layers=show:0&bbox=-18.57711821048096,49.844261145603,12.61594729098715,58.66349271362658&bboxSR=4326&imageSR=27700&size=1680,914&f=image](https://map.bgs.ac.uk/arcgis/rest/services/GeoIndex_Onshore/hydrogeology/MapServer//export?dpi=96&transparent=true&format=png&layers=show:0&bbox=-18.57711821048096,49.844261145603,12.61594729098715,58.66349271362658&bboxSR=4326&imageSR=27700&size=1680,914&f=image)

Now this produces the image like so: 
<img width="1056" alt="Screenshot 2021-11-30 at 18 49 27" src="https://user-images.githubusercontent.com/49888000/144109350-e77895fb-d2cc-4e58-9a45-51a17bfd91c5.png">

Now i think my issue seems to be the fact the image "tile" that is loaded here is not a square, it is a 599x914 image (based on viewport size). The code to generate this is here: 
```
var baseUrl = `https://map.bgs.ac.uk/arcgis/rest/services/GeoIndex_Onshore/hydrogeology/MapServer/`
    const wmsLayer = new TileLayer({
      id: 'wmsLayer',
      data: baseUrl,
      getTileData: (tile) => {
        console.log("🚀 ~ file: wmsLayer.js ~ line 30 ~ tile", tile)
        var urlToModify = `${baseUrl}/export?dpi=96&transparent=true&format=png&layers=show:0&bbox=${latlngArray[0]},${latlngArray[1]},${latlngArray[2]},${latlngArray[3]}&bboxSR=4326&imageSR=27700&size=${heightAndWidth[1]},${heightAndWidth[0]}&f=image`
        return urlToModify
      },
      minZoom: 2,
      transparentColor: [255, 255, 255, 0],
      colorFormat: 'RGBA',
      maxZoom: 18,
      tileSize: heightAndWidth[0],
      updateTriggers: {
        visible: layerVisibility,
        opacity: layerOpacity,
        getTileData: latlngArray
      },
      visible: layerVisibility.wmsLayer,
      opacity: layerOpacity.wmsLayer,
      zoomOffset: devicePixelRatio === 1 ? -1 : 0,
      onDataLoad: (value, context) => {
        console.log("Data loaded", value, context)
      },
      // extent: [-14.888427318, 63.888268748, 47.968268748, 3.399072682],
      renderSubLayers: (props) => {
        const {
          bbox: {west, south, east, north}
        } = props.tile;
        return new BitmapLayer(props, {
          data: null,
          image: props.data,
          opacity: layerOpacity.wmsLayer,
          transparentColor: [255, 255, 255, 0],
          bounds: [latlngArray[0], latlngArray[1], latlngArray[2], latlngArray[3]],
          updateTriggers: {
            visible: layerVisibility,
            bounds: latlngArray,
            opacity: layerOpacity,
            // image: latlngArray
          },
      })}
      
    })
    return wmsLayer
```

Latlng Array is the viewport projected to mercator like so: 
```
          const vp = new WebMercatorViewport(e.viewState);

          const nw = vp.viewport.unproject([0, 0]);
          const se = vp.viewport.unproject([
                     vp.viewport.width,
                      vp.viewport.height,
          ]);
```
This is used to return a bounding box array like so: 
```
[-15.374258806312191,44.33316121698344,-4.252445570967491,55.228263885624486]
```

heightAndWidth is the viewport height and width
```
[599,914]
```
So I am wondering, is this an issue with the image size that it is not "fitting" into the tile layer as the tile layer is a square so the offset comes from here, or another projection related issue. Any points towards the correct direction would be greatly appreciated

**Comment by kylebarron** [comment, marked as answer]
> `imageSR=27700`

It looks like that's telling GeoServer to load the image in a [UK-specific projection](https://epsg.io/27700). You should try to load it in Web Mercator https://epsg.io/3857

**Reply by lm93547** [reply]
This worked perfectly. Thanks
